INT-319: Introduce Drag and Drop to Embed

kirisanth-g opened the pull request:
### What
* Introduce `dragAndDrop` option to embed
* If turned on Button & Intros will be draggable
* Button/Intro will not active during drag
### Test
* Load embed and drag it around
* Open & Close Chat should still will be where it was left
---
- [x] PR title follows the format \<Jira Card\>: \<Change description\>
- [x] PR description present and accurate
- [ ] PR is linked to a Jira card has `impact` field filled in
- [x] Unit tests cover code changes
- [ ] Changes have gone through design review if required
- [ ] Deployment plan and/or bridge code description is included in PR if required
- [x] Changes are CATO/WCAG compliant and support IE9+
- [ ] A [Release Note](https://www.notion.so/adasupport/Creating-Release-Notes-36906dfa8a2b4f10a31dc23b8d46681a) has been drafted and published after merging to master

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,180 @@
+    } else if (!dragging && prevState.dragging) {
+      document.removeEventListener("mousemove", this.onMouseMove);
+      document.removeEventListener("mouseup", this.onUp);
+      document.addEventListener("touchmove", this.onTouchMove);
```
These should be be `removeEventListener`, no?

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,180 @@
+
+  delayedSetter() {
+    setTimeout(
+      (() => {
```
Looks like you have some extra brackets happening, can't you just do:

```
setTimeout(() => {
  this.setState({ beingDragged: false });
}, 250);
```
?

NicHaley replied:
For that matter, do we need both `beingDragged` and `dragging`? What would happen if we just used `dragging`?

kirisanth-g replied:
Extra brackets are for the binding, and we need beingDragged to be delayed compared to dragging as events as all the mouseUp, touchEnd events become active again otherwise

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,180 @@
+    e.preventDefault();
+  }
+
+  onStart(e: MouseEvent | TouchEventItem) {
```
Let's give this a more explicit name

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,180 @@
+  }
+
+  onMove(e: MouseEvent | TouchEventItem) {
+    if (!this.state.dragging) return;
```
Use object destructuring

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,180 @@
+  onMove(e: MouseEvent | TouchEventItem) {
+    if (!this.state.dragging) return;
+    const { updatePosition } = this.props;
+    const x = Math.max(window.innerWidth - e.pageX + this.state.rel.x, 0);
```
Same here

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,180 @@
+    const { pos, beingDragged } = this.state;
+    const { children } = this.props;
+    const divStyle = {
+      position:"fixed",
```
Missing space

NicHaley commented on src/components/App/index.tsx:
```diff
@@ -84,7 +87,8 @@ export default class App extends Component<InterfaceApp, InterfaceState> {
       unreadMessages: 0,
-      hasConnectedChat: false
+      hasConnectedChat: false,
+      buttonPosition: { x: 24, y: 24 }
```
Add these values to constants file

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,180 @@
+        this.setState({ beingDragged: false });
+      })
+      .bind(this),
+      250
```
What are the downsides of making this number smaller? When testing I dragged a clicked immediately after and it's  a little weird when it doesn't open on click

kirisanth-g replied:
Lowering the speed to 50

kirisanth-g commented:
> Dude this is super sick!! I added a few comments inline. Also I had some other questions/feedback:
> 
> 1. The current behaviour for the chat button is to hide when the drawer is opened. This was fine before because the button was behind the drawer, but now it looks weird because the button visibly disappears when the button is moved and drawer opened. We should change this behaviour. (I'll send you a video of this)
> 2. At one point our designs featured button snapping into the corners. Is this still part of the design?
> 3. Not sure if you've done this but be sure to do thorough testing on mobile and on IE! :)

1. Fixed! If `dragAndDrop` is true button will stay rendered
2. We decided moving anywhere was much more appealing of a feature than snapping to only corners
3. Will be doing that soon!

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,209 @@
+
+interface DraggableState {
+  dragging: boolean,
+  rel: {x: number, y: number},
```
Looks like our linter hasn't been configured for this yet, but be sure to add a space on either side of the curly brackets (right now it's inconsistent)

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,209 @@
+      document.addEventListener("mouseup", this.onUp);
+      document.addEventListener("touchmove", this.onTouchMove, { passive: false });
+      document.addEventListener("touchend", this.onUp);
+      document.addEventListener("click", this.handleOnClick, true);
```
Do we need to remove this handler when not dragging?

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,209 @@
+    e.stopPropagation();
+
+    // iOS fix
+    if (/iPad|iPhone|iPod/.test(navigator.userAgent)) {
```
We should add this check to the `browsers.ts` service file

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,209 @@
+      );
+    e.stopPropagation();
+
+    // iOS fix
```
This could use more of a description. Why does iOS need an artificial MouseEvent?

NicHaley commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,209 @@
+      {children}
+    </Draggable>
+  )
+  : <div> {children} </div>
```
No need for spaces inside of div

NicHaley commented on src/components/App/IntroBlurb/index.tsx:
```diff
@@ -4,10 +4,13 @@ import { Component, h } from "preact";
   client: Client,
   isInMobile: boolean,
+  isDraggable: boolean,
+  dragPosition?: {x: number, y: number},
```
Watch spacing

agreenbloom commented on src/components/App/Draggable/index.tsx:
```diff
@@ -0,0 +1,213 @@
+  };
+  myRef: any;
+
+  constructor(props: DraggableProps) {
```
just curious why you choose the constructor approach over setting `state={..}` at the top of the class and binding functions with the `=>`
